src/android/com/microsoftopentechnologies/intellij/wizards/createvm/CreateCloudServiceForm.java

```diff
@@ -0,0 +1,217 @@
+            public void actionPerformed(ActionEvent e) {
+                onCancel();
+            }
+        });
+
+        setDefaultCloseOperation(DISPOSE_ON_CLOSE);
+        addWindowListener(new WindowAdapter() {
+            public void windowClosing(WindowEvent e) {
+                onCancel();
+            }
+        });
+
+        contentPane.registerKeyboardAction(new ActionListener() {
+            public void actionPerformed(ActionEvent e) {
+                onCancel();
+            }
+        }, KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0), JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT);
+
+        regionOrAffinityGroupComboBox.setRenderer(new DefaultListCellRenderer() {
+            @Override
+            public Component getListCellRendererComponent(JList<?> jList, Object o, int i, boolean b, boolean b1) {
+
+                return (o instanceof String) ?
+                        super.getListCellRendererComponent(jList, o, i, b, b1)
+                        : super.getListCellRendererComponent(jList, "  " + o.toString(), i, b, b1);
+            }
+        });
+
+        nameTextField.getDocument().addDocumentListener(new DocumentListener() {
+            @Override
+            public void insertUpdate(DocumentEvent documentEvent) {
+                validateEmptyFields();
+            }
+
+            @Override
+            public void removeUpdate(DocumentEvent documentEvent) {
+                validateEmptyFields();
+            }
+
+            @Override
+            public void changedUpdate(DocumentEvent documentEvent) {
+                validateEmptyFields();
+            }
+        });
+
+        regionOrAffinityGroupComboBox.addItemListener(new ItemListener() {
+            @Override
+            public void itemStateChanged(ItemEvent itemEvent) {
+                validateEmptyFields();
+            }
+        });
+
+    }
+
+    private void validateEmptyFields() {
+        boolean allFieldsCompleted = !(
+                nameTextField.getText().isEmpty() || regionOrAffinityGroupComboBox.getSelectedObjects().length == 0);
+
+        buttonOK.setEnabled(allFieldsCompleted);
+    }
+
+    public void fillFields(final Subscription subscription, Project project) {
+        this.subscription = subscription;
+
+        subscriptionComboBox.addItem(subscription.getName());
+
+        regionOrAffinityGroupComboBox.addItem("<Loading...>");
+
+        ProgressManager.getInstance().run(new Task.Backgroundable(project, "Loading regions...", false) {
+            @Override
+            public void run(ProgressIndicator progressIndicator) {
+                progressIndicator.setIndeterminate(true);
+
+                try {
+
+                    final List<AffinityGroup> affinityGroups = AzureSDKManagerImpl.getManager().getAffinityGroups(subscription.getId().toString());
+                    final List<Location> locations = AzureSDKManagerImpl.getManager().getLocations(subscription.getId().toString());
+
```

Comment:
We need to figure out a way of efficiently running multiple async I/O calls such as the two lines above in parallel. There's no reason to wait till the affinity groups are loaded before issuing the call to fetch the list of locations. Both can run in parallel and we should be able to wait on the result of both the operations to complete before continuing.